File: Android/appdimens_dynamic/src/main/java/com/appdimens/dynamic/compose/AppDimensDynamic.kt
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/bodenberg/appdimens.git
 * Date: 2025-10-04
 *
 * Library: AppDimens
 *
 * Description:
 * The AppDimens library is a dimension management system that automatically
 * adjusts Dp, Sp, and Px values in a responsive and mathematically refined way,
 * ensuring layout consistency across any screen size or ratio.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.appdimens.dynamic.compose

import android.annotation.SuppressLint
import android.content.res.Configuration
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.appdimens.dynamic.compose.AppDimensAdjustmentFactors.BASE_WIDTH_DP
import com.appdimens.dynamic.compose.AppDimensAdjustmentFactors.resolveIntersectionCondition
import com.appdimens.dynamic.compose.AppDimensAdjustmentFactors.resolveQualifierDp
import com.appdimens.library.DpQualifier
import com.appdimens.library.DpQualifierEntry
import com.appdimens.library.ScreenType
import com.appdimens.library.UiModeQualifierEntry
import com.appdimens.library.UiModeType

/**
 * [EN] A builder class for creating dynamic dimensions that allow base Dp customization
 * via screen qualifiers (`.screen()`). The final value is scaled by the screen size.
 *
 * [PT] Uma classe construtora para criar dimensões dinâmicas que permitem a customização do Dp base
 * por meio de qualificadores de tela (`.screen()`). O valor final é escalado pelo tamanho da tela.
 */
class AppDimensDynamic(
    private val initialBaseDp: Dp,
    private var ignoreMultiViewAdjustment: Boolean = false
) {
    
    /**
     * [EN] Map to store custom Dp values (Priority 3).
     *
     * [PT] Mapa para armazenar valores Dp customizados (Prioridade 3).
     */
    private var customDpMap: MutableMap<DpQualifierEntry, Dp> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by UiModeType (Priority 2).
     *
     * [PT] Mapa para valores Dp customizados por UiModeType (Prioridade 2).
     */
    private var customUiModeMap: MutableMap<UiModeType, Dp> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by INTERSECTION (UiMode + DpQualifier) (Priority 1).
     *
     * [PT] Mapa para valores Dp customizados por INTERSEÇÃO (UiMode + DpQualifier) (Prioridade 1).
     */
    private var customIntersectionMap: MutableMap<UiModeQualifierEntry, Dp> = mutableMapOf()

    private var screenType: ScreenType = ScreenType.LOWEST
    
    /**
     * [EN] Individual remember control for this instance.
     * [PT] Controle individual de remember para esta instância.
     */
    private var enableRemember: Boolean = true

    /**
     * [EN] Sets a custom dimension for a specific UI mode.
     *
     * [PT] Define uma dimensão customizada para um modo de UI específico.
     */
    fun screen(type: UiModeType, customValue: Dp): AppDimensDynamic {
        customUiModeMap[type] = customValue
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a TextUnit value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor TextUnit.
     */
    fun screen(type: UiModeType, customValue: TextUnit): AppDimensDynamic {
        customUiModeMap[type] = customValue.value.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a Float value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Float.
     */
    fun screen(type: UiModeType, customValue: Float): AppDimensDynamic {
        customUiModeMap[type] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts an Int value.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Int.
     */
    fun screen(type: UiModeType, customValue: Int): AppDimensDynamic {
        customUiModeMap[type] = customValue.dp
        return this
    }

    /**
     * [EN] Sets a custom dimension for a specific intersection of UI mode and screen qualifier.
     *
     * [PT] Define uma dimensão customizada para uma interseção específica de modo de UI e qualificador de tela.
     */
    fun screen(
        uiModeType: UiModeType, qualifierType: DpQualifier, qualifierValue: Int, customValue: Dp
    ): AppDimensDynamic {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts a TextUnit value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor TextUnit.
     */
    fun screen(uiModeType: UiModeType, qualifierType: DpQualifier, qualifierValue: Int, customValue: TextUnit): AppDimensDynamic {
        val key = UiModeQualifierEntry(uiModeType = uiModeType, dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue))
        customIntersectionMap[key] = customValue.value.dp
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts a Float value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor Float.
     */
    fun screen(uiModeType: UiModeType, qualifierType: DpQualifier, qualifierValue: Int, customValue: Float): AppDimensDynamic {
        val key = UiModeQualifierEntry(uiModeType = uiModeType, dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue))
        customIntersectionMap[key] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` intersection that accepts an Int value.
     *
     * [PT] Sobrecarga para a interseção `screen` que aceita um valor Int.
     */
    fun screen(uiModeType: UiModeType, qualifierType: DpQualifier, qualifierValue: Int, customValue: Int): AppDimensDynamic {
        val key = UiModeQualifierEntry(uiModeType = uiModeType, dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue))
        customIntersectionMap[key] = customValue.dp
        return this
    }

    /**
     * [EN] Sets a custom dimension for a specific screen qualifier.
     *
     * [PT] Define uma dimensão customizada para um qualificador de tela específico.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Int): AppDimensDynamic {
        customDpMap[DpQualifierEntry(type, value)] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a Float value for a DpQualifier.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor Float para um DpQualifier.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Float): AppDimensDynamic {
        customDpMap[DpQualifierEntry(type, value)] = customValue.dp
        return this
    }

    /**
     * [EN] Overload for `screen` that accepts a TextUnit value for a DpQualifier.
     *
     * [PT] Sobrecarga para `screen` que aceita um valor TextUnit para um DpQualifier.
     */
    fun screen(type: DpQualifier, value: Int, customValue: TextUnit): AppDimensDynamic {
        customDpMap[DpQualifierEntry(type, value)] = customValue.value.dp
        return this
    }

    /**
     * [EN] Sets the screen dimension type to be used as a reference (HIGHEST or LOWEST).
     *
     * [PT] Define o tipo de dimensão da tela a ser usado como referência (HIGHEST ou LOWEST).
     */
    fun type(type: ScreenType): AppDimensDynamic {
        screenType = type
        return this
    }

    /**
     * [EN] Ignores multi-view adjustment if set to true.
     *
     * [PT] Ignora o ajuste de multi-view se definido como verdadeiro.
     */
    fun multiViewAdjustment(ignore: Boolean = true): AppDimensDynamic {
        ignoreMultiViewAdjustment = ignore
        return this
    }
    
    /**
     * [EN] Enables or disables remember for this instance.
     * @param enable If true, enables remember; if false, disables remember.
     * @return The AppDimensDynamic instance for chaining.
     * [PT] Ativa ou desativa o remember para esta instância.
     * @param enable Se verdadeiro, ativa o remember; se falso, desativa o remember.
     * @return A instância AppDimensDynamic para encadeamento.
     */
    fun remember(enable: Boolean = true): AppDimensDynamic {
        enableRemember = enable
        return this
    }

    @SuppressLint("ConfigurationScreenWidthHeight")
    @Composable
    private fun rememberFinalBaseDp(): Dp {
        val configuration = LocalConfiguration.current

        val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
        val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
        val currentScreenHeightDp = configuration.screenHeightDp.toFloat()

        val currentUiModeType = UiModeType.fromConfiguration(configuration.uiMode)

        return if (AppDimens.globalRememberEnabled && enableRemember) {
            remember(
                smallestWidthDp, currentScreenWidthDp, currentScreenHeightDp,
                customDpMap.hashCode(), customUiModeMap.hashCode(), customIntersectionMap.hashCode(),
                currentUiModeType
            ) {
                calculateBaseDp(
                    smallestWidthDp, currentScreenWidthDp, currentScreenHeightDp,
                    currentUiModeType
                )
            }
        } else {
            calculateBaseDp(
                smallestWidthDp, currentScreenWidthDp, currentScreenHeightDp,
                currentUiModeType
            )
        }
    }
    
    @SuppressLint("ConfigurationScreenWidthHeight")
    private fun calculateBaseDp(
        smallestWidthDp: Float,
        currentScreenWidthDp: Float,
        currentScreenHeightDp: Float,
        currentUiModeType: UiModeType
    ): Dp {
            var dpToAdjust = initialBaseDp
            var foundCustomDp: Dp?

            /**
             * [EN] PRIORITY 1: INTERSECTION (UiMode + DpQualifier)
             *
             * [PT] PRIORIDADE 1: INTERSEÇÃO (UiMode + DpQualifier)
             */
            val sortedIntersectionQualifiers = customIntersectionMap.entries.toList()
                .sortedByDescending { it.key.dpQualifierEntry.value }

            foundCustomDp = sortedIntersectionQualifiers.firstOrNull { (key, _) ->
                key.uiModeType == currentUiModeType && resolveIntersectionCondition(
                    entry = key.dpQualifierEntry,
                    smallestWidthDp = smallestWidthDp,
                    currentScreenWidthDp = currentScreenWidthDp,
                    currentScreenHeightDp = currentScreenHeightDp
                )
            }?.value

            if (foundCustomDp != null) {
                dpToAdjust = foundCustomDp
            } else {
                /**
                 * [EN] PRIORITY 2: UI MODE (UiModeType only)
                 *
                 * [PT] PRIORIDADE 2: UI MODE (Apenas UiModeType)
                 */
                foundCustomDp = customUiModeMap[currentUiModeType]

                if (foundCustomDp != null) {
                    dpToAdjust = foundCustomDp
                } else {
                    /**
                     * [EN] PRIORITY 3: DP QUALIFIER (SW, H, W only)
                     *
                     * [PT] PRIORIDADE 3: DP QUALIFIER (Apenas SW, H, W)
                     */
                    dpToAdjust = resolveQualifierDp(
                        customDpMap = customDpMap,
                        smallestWidthDp = smallestWidthDp,
                        currentScreenWidthDp = currentScreenWidthDp,
                        currentScreenHeightDp = currentScreenHeightDp,
                        initialBaseDp = initialBaseDp
                    )
                }
            }
            return dpToAdjust
    }


    @SuppressLint("ConfigurationScreenWidthHeight")
    @Composable
    private fun calculate(): Float {
        val dpToAdjust = rememberFinalBaseDp()
        val configuration = LocalConfiguration.current

        var isMultiWindow = ignoreMultiViewAdjustment

        if (ignoreMultiViewAdjustment) {
            val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
            val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
            val isLayoutSplit = configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK != Configuration.SCREENLAYOUT_SIZE_MASK
            val isSmallDifference = (smallestWidthDp - currentScreenWidthDp) < (smallestWidthDp * 0.1)
            isMultiWindow = isLayoutSplit && !isSmallDifference
        }

        val shouldIgnoreAdjustment = ignoreMultiViewAdjustment && isMultiWindow

        if (shouldIgnoreAdjustment) {
            /**
             * [EN] Returns the base Dp without dynamic scaling.
             *
             * [PT] Retorna o Dp base sem escalonamento dinâmico.
             */
            return dpToAdjust.value
        }

        /**
         * [EN] The dynamic scaling percentage is: (Adjusted Base DP / Reference DP).
         *
         * [PT] A porcentagem de escalonamento dinâmico é: (DP Base Ajustado / DP de Referência).
         */
        val percentage = dpToAdjust.value / BASE_WIDTH_DP

        /**
         * [EN] Screen dimension to use (LOWEST or HIGHEST).
         *
         * [PT] Dimensão da tela a ser usada (LOWEST ou HIGHEST).
         */
        val dimensionToUse = when (screenType) {
            ScreenType.HIGHEST -> maxOf(configuration.screenWidthDp.toFloat(), configuration.screenHeightDp.toFloat())
            ScreenType.LOWEST -> minOf(configuration.screenWidthDp.toFloat(), configuration.screenHeightDp.toFloat())
        }

        /**
         * [EN] The final value is the percentage applied to the screen dimension.
         *
         * [PT] O valor final é a porcentagem aplicada à dimensão da tela.
         */
        return dimensionToUse * percentage
    }

    /**
     * [EN] Builds the adjusted Dp from the calculation.
     *
     * [PT] Constrói o Dp ajustado a partir do cálculo.
     */
    @get:Composable
    val dp: Dp
        get() = calculate().dp

    /**
     * [EN] Builds the adjusted TextUnit (Sp) from the calculation.
     *
     * [PT] Constrói o TextUnit (Sp) ajustado a partir do cálculo.
     */
    @get:Composable
    val sp: TextUnit
        get() = calculate().sp

    /**
     * [EN] Builds the adjusted TextUnit (Sp) from the calculation (NO FONT SCALE).
     *
     * [PT] Constrói o TextUnit (Sp) ajustado a partir do cálculo (SEM ESCALA DE FONTE).
     */
    @get:Composable
    val em: TextUnit
        get() = (calculate() / LocalDensity.current.fontScale).sp

    /**
     * [EN] Builds the adjusted Pixel value (Float) from the calculation.
     *
     * [PT] Constrói o valor em Pixels (Float) ajustado a partir do cálculo.
     */
    @get:Composable
    val px: Float
        get() = calculate()
}

File: ReactNative/src/core/AppDimensDynamic.ts
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/bodenberg/appdimens.git
 * Date: 2025-01-27
 *
 * Library: AppDimens React Native - Dynamic Scaling
 *
 * Description:
 * Dynamic dimension scaling with proportional adjustment for aggressive scaling.
 * Ideal for layout containers and fluid elements.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  DimensionConfig,
  AppDimensBuilder,
  UiModeType,
  DeviceType,
  ScreenQualifier,
  ScreenQualifierEntry,
  UiModeQualifierEntry,
  BASE_WIDTH_DP,
} from '../types';
import {AppDimensAdjustmentFactors} from './AppDimensAdjustmentFactors';
import {resolveScreenType} from '../utils/orientationResolver';
import type {BaseOrientation} from '../types/BaseOrientation';

/**
 * Dynamic dimension builder with proportional adjustment
 */
export class AppDimensDynamic implements AppDimensBuilder {
  private config: DimensionConfig;
  private baseValueCache = new Map<string, number>();
  private calculatedValueCache = new Map<string, number>();

  constructor(
    initialValue: number,
    ignoreMultiWindowAdjustment: boolean = false,
  ) {
    this.config = {
      initialValue,
      screenType: 'lowest',
      baseOrientation: 'auto',
      applyAspectRatioAdjustment: false, // Dynamic doesn't use aspect ratio
      ignoreMultiWindowAdjustment,
      customScreenQualifierMap: new Map(),
      customUiModeMap: new Map(),
      customIntersectionMap: new Map(),
    };
  }

  /**
   * [EN] Set custom value for UI mode.
   * Compatible with Android/iOS API.
   * [PT] Define valor customizado para modo UI.
   * Compatível com API Android/iOS.
   */
  screen(uiModeType: UiModeType, customValue: number): this;

  /**
   * [EN] Set custom value for device type and screen size.
   * Compatible with Android/iOS API.
   * [PT] Define valor customizado para tipo de dispositivo e tamanho de tela.
   * Compatível com API Android/iOS.
   */
  screen(deviceType: DeviceType, screenSize: number, customValue: number): this;

  /**
   * [EN] Set custom value for screen qualifier with specific value.
   * Compatible with Android/iOS API.
   * [PT] Define valor customizado para qualificador de tela com valor específico.
   * Compatível com API Android/iOS.
   */
  screen(
    qualifierType: ScreenQualifier,
    qualifierValue: number,
    customValue: number,
  ): this;

  /**
   * [EN] Set custom value for intersection of UI mode and screen qualifier.
   * Compatible with Android/iOS API.
   * [PT] Define valor customizado para interseção de modo UI e qualificador de tela.
   * Compatível com API Android/iOS.
   */
  screen(
    uiModeType: UiModeType,
    qualifierType: ScreenQualifier,
    qualifierValue: number,
    customValue: number,
  ): this;

  /**
   * Set screen dimension type
   * Primary method - matches Android, iOS, Flutter API
   */
  screen(type: 'lowest' | 'highest'): this;

  // Implementation
  screen(
    arg1: UiModeType | DeviceType | ScreenQualifier | 'lowest' | 'highest',
    arg2?: number | ScreenQualifier,
    arg3?: number,
    arg4?: number,
  ): this {
    // Check if it's screen(type: 'lowest' | 'highest')
    if (arg1 === 'lowest' || arg1 === 'highest') {
      this.config.screenType = arg1;
      this.invalidateCache();
      return this;
    }

    if (arg3 === undefined && arg4 === undefined) {
      // screen(uiModeType, customValue)
      const uiModeType = arg1 as UiModeType;
      const customValue = arg2 as number;
      this.config.customUiModeMap.set(uiModeType, customValue);
    } else if (arg4 === undefined && arg3 !== undefined) {
      // Could be: screen(deviceType, screenSize, customValue) or
      //           screen(qualifierType, qualifierValue, customValue)
      const isDeviceType = [
        'phone',
        'tablet',
        'desktop',
        'watch',
        'tv',
        'car',
      ].includes(arg1);

      if (isDeviceType) {
        // screen(deviceType, screenSize, customValue)
        const deviceType = arg1 as DeviceType;
        const screenSize = arg2 as number;
        const customValue = arg3;
        const screenQualifier: ScreenQualifierEntry = {
          type: 'smallWidth',
          value: screenSize,
          deviceType,
        };
        this.config.customScreenQualifierMap.set(screenQualifier, customValue);
      } else {
        // screen(qualifierType, qualifierValue, customValue)
        const qualifierType = arg1 as ScreenQualifier;
        const qualifierValue = arg2 as number;
        const customValue = arg3;
        const screenQualifier: ScreenQualifierEntry = {
          type: qualifierType,
          value: qualifierValue,
        };
        this.config.customScreenQualifierMap.set(screenQualifier, customValue);
      }
    } else if (arg4 !== undefined && arg3 !== undefined) {
      // screen(uiModeType, qualifierType, qualifierValue, customValue)
      const uiModeType = arg1 as UiModeType;
      const qualifierType = arg2 as ScreenQualifier;
      const qualifierValue = arg3;
      const customValue = arg4;
      const screenQualifier: ScreenQualifierEntry = {
        type: qualifierType,
        value: qualifierValue,
        uiMode: uiModeType,
      };
      const intersectionEntry: UiModeQualifierEntry = {
        uiModeType,
        screenQualifierEntry: screenQualifier,
      };
      this.config.customIntersectionMap.set(intersectionEntry, customValue);
    }
    this.invalidateCache();
    return this;
  }

  /**
   * Alias for screen() - kept for backward compatibility
   * @deprecated Use screen() instead for consistency with other platforms
   */
  type(type: 'lowest' | 'highest'): this {
    return this.screen(type);
  }

  baseOrientation(orientation: 'portrait' | 'landscape' | 'auto'): this {
    this.config.baseOrientation = orientation;
    return this;
  }

  portraitLowest(): this {
    this.config.baseOrientation = 'portrait';
    this.config.screenType = 'lowest';
    return this;
  }

  portraitHighest(): this {
    this.config.baseOrientation = 'portrait';
    this.config.screenType = 'highest';
    return this;
  }

  landscapeLowest(): this {
    this.config.baseOrientation = 'landscape';
    this.config.screenType = 'lowest';
    return this;
  }

  landscapeHighest(): this {
    this.config.baseOrientation = 'landscape';
    this.config.screenType = 'highest';
    return this;
  }

  /**
   * Set multi-window adjustment behavior
   */
  multiWindowAdjustment(ignore: boolean = true): this {
    this.config.ignoreMultiWindowAdjustment = ignore;
    this.invalidateCache();
    return this;
  }

  /**
   * Enable/disable aspect ratio adjustment (not used in dynamic)
   */
  aspectRatio(enable: boolean = true, sensitivity?: number): this {
    // Dynamic scaling doesn't use aspect ratio adjustment
    return this;
  }

  /**
   * Enable/disable cache
   */
  cache(enable: boolean = true): this {
    if (!enable) {
      this.invalidateCache();
    }
    return this;
  }

  /**
   * Calculate adjusted value in pixels
   */
  toPixels(): number {
    const adjustedValue = this.calculateAdjustedValue();
    return AppDimensAdjustmentFactors.pointsToPixels(adjustedValue);
  }

  /**
   * Calculate adjusted value in pixels (integer)
   */
  toPixelsInt(): number {
    return Math.round(this.toPixels());
  }

  /**
   * Calculate adjusted value in points
   */
  toPoints(): number {
    return this.calculateAdjustedValue();
  }

  /**
   * Calculate adjusted value in points (integer)
   */
  toPointsInt(): number {
    return Math.round(this.calculateAdjustedValue());
  }

  /**
   * [EN] Calculate adjusted value in scalable pixels (sp) for text.
   * This respects the system font scale setting.
   * Compatible with Android/iOS API.
   * [PT] Calcula o valor ajustado em pixels escaláveis (sp) para texto.
   * Isso respeita a configuração de escala de fonte do sistema.
   * Compatível com API Android/iOS.
   */
  toSp(): number {
    const adjustedValue = this.calculateAdjustedValue();
    const fontScale = AppDimensAdjustmentFactors.getFontScale();
    return adjustedValue * fontScale;
  }

  /**
   * [EN] Calculate adjusted value in scalable pixels (sp) as integer.
   * Compatible with Android/iOS API.
   * [PT] Calcula o valor ajustado em pixels escaláveis (sp) como inteiro.
   * Compatível com API Android/iOS.
   */
  toSpInt(): number {
    return Math.round(this.toSp());
  }

  /**
   * [EN] Calculate adjusted value in em units (ignoring font scale).
   * This is similar to sp but without the font scale factor applied.
   * Compatible with Android/iOS API.
   * [PT] Calcula o valor ajustado em unidades em (ignorando escala de fonte).
   * Isso é similar ao sp mas sem o fator de escala de fonte aplicado.
   * Compatível com API Android/iOS.
   */
  toEm(): number {
    return this.calculateAdjustedValue();
  }

  /**
   * [EN] Calculate adjusted value in em units as integer.
   * Compatible with Android/iOS API.
   * [PT] Calcula o valor ajustado em unidades em como inteiro.
   * Compatível com API Android/iOS.
   */
  toEmInt(): number {
    return Math.round(this.toEm());
  }

  /**
   * Resolve final base value using customization logic
   */
  private resolveFinalBaseValue(): number {
    const screenDimensions =
      AppDimensAdjustmentFactors.getCurrentScreenDimensions();
    const deviceInfo = AppDimensAdjustmentFactors.getCurrentDeviceInfo();
    const {width, height} = screenDimensions;
    const currentScreenSize = Math.min(width, height);

    // Generate cache key
    const cacheKey = `${deviceInfo.type}_${deviceInfo.platform}_${width}_${height}_${this.config.initialValue}_${this.config.customIntersectionMap.size}_${this.config.customUiModeMap.size}_${this.config.customScreenQualifierMap.size}`;

    // Check cache first
    if (this.baseValueCache.has(cacheKey)) {
      return this.baseValueCache.get(cacheKey)!;
    }

    let valueToAdjust = this.config.initialValue;

    // Priority 1: Intersection (UiMode + ScreenQualifier)
    const sortedIntersectionQualifiers = Array.from(
      this.config.customIntersectionMap.keys(),
    ).sort(
      (a, b) => b.screenQualifierEntry.value - a.screenQualifierEntry.value,
    );

    for (const key of sortedIntersectionQualifiers) {
      if (
        key.uiModeType === this.getCurrentUiMode() &&
        AppDimensAdjustmentFactors.resolveIntersectionCondition(
          key.screenQualifierEntry,
          currentScreenSize,
          width,
          height,
        )
      ) {
        valueToAdjust =
          this.config.customIntersectionMap.get(key) ||
          this.config.initialValue;
        break;
      }
    }

    // Priority 2: UI Mode
    if (valueToAdjust === this.config.initialValue) {
      const uiModeValue = this.config.customUiModeMap.get(
        this.getCurrentUiMode(),
      );
      if (uiModeValue !== undefined) {
        valueToAdjust = uiModeValue;
      }
    }

    // Priority 3: Device Type / Screen Qualifier
    if (valueToAdjust === this.config.initialValue) {
      valueToAdjust = AppDimensAdjustmentFactors.resolveQualifierValue(
        this.config.customScreenQualifierMap,
        deviceInfo.type,
        currentScreenSize,
        this.config.initialValue,
      );
    }

    // Cache the result
    this.baseValueCache.set(cacheKey, valueToAdjust);

    return valueToAdjust;
  }

  /**
   * Calculate final adjusted value using dynamic scaling
   */
  private calculateAdjustedValue(): number {
    const screenDimensions =
      AppDimensAdjustmentFactors.getCurrentScreenDimensions();
    const {width, height} = screenDimensions;

    // Generate cache key for final calculation
    const cacheKey = `${this.config.screenType}_${this.config.ignoreMultiWindowAdjustment}_${width}_${height}_${this.config.initialValue}_${this.config.customIntersectionMap.size}_${this.config.customUiModeMap.size}_${this.config.customScreenQualifierMap.size}`;

    // Check cache first
    if (this.calculatedValueCache.has(cacheKey)) {
      return this.calculatedValueCache.get(cacheKey)!;
    }

    const valueToAdjust = this.resolveFinalBaseValue();

    const shouldIgnoreAdjustment =
      this.config.ignoreMultiWindowAdjustment &&
      AppDimensAdjustmentFactors.isMultiWindowMode();

    let finalValue: number;

    if (shouldIgnoreAdjustment) {
      // Return base value without dynamic scaling
      finalValue = valueToAdjust;
    } else {
      // Dynamic scaling: percentage of screen dimension
      const percentage = valueToAdjust / BASE_WIDTH_DP;

      // Resolve effective screen type based on base orientation
      const effectiveScreenType = resolveScreenType(
        this.config.screenType,
        this.config.baseOrientation || 'auto',
        {width, height},
      );

      // Screen dimension to use (LOWEST or HIGHEST)
      const dimensionToUse =
        effectiveScreenType === 'highest'
          ? Math.max(width, height)
          : Math.min(width, height);

      // Final value is percentage applied to screen dimension
      finalValue = dimensionToUse * percentage;
    }

    // Cache the result
    this.calculatedValueCache.set(cacheKey, finalValue);

    return finalValue;
  }

  /**
   * Get current UI mode
   */
  private getCurrentUiMode(): UiModeType {
    // In React Native, we can't easily detect UI mode
    // This is a placeholder for future implementation
    return 'normal';
  }

  /**
   * Invalidate cache
   */
  private invalidateCache(): void {
    this.baseValueCache.clear();
    this.calculatedValueCache.clear();
  }
}
